File: packages/ruby-sdk/lib/strait/operations/workflow_runs.rb
# frozen_string_literal: true

module Strait
  module Operations
    # Service for workflow run lifecycle management and step-level operations.
    class WorkflowRunsService < BaseService
      # List all workflow runs.
      # GET /v1/workflow-runs
      def list(query: nil)
        _request(:get, "/v1/workflow-runs", query: query)
      end

      # Get a workflow run by ID.
      # GET /v1/workflow-runs/{workflowRunID}
      def get(workflow_run_id)
        _request(:get, "/v1/workflow-runs/{workflowRunID}", path_params: { "workflowRunID" => workflow_run_id })
      end

      # Delete a workflow run.
      # DELETE /v1/workflow-runs/{workflowRunID}
      def delete(workflow_run_id)
        _request(:delete, "/v1/workflow-runs/{workflowRunID}", path_params: { "workflowRunID" => workflow_run_id })
        nil
      end

      # Pause a workflow run.
      # POST /v1/workflow-runs/{workflowRunID}/pause
      def pause(workflow_run_id)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/pause", path_params: { "workflowRunID" => workflow_run_id })
      end

      # Resume a paused workflow run.
      # POST /v1/workflow-runs/{workflowRunID}/resume
      def resume(workflow_run_id)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/resume", path_params: { "workflowRunID" => workflow_run_id })
      end

      # Retry a failed workflow run.
      # POST /v1/workflow-runs/{workflowRunID}/retry
      def retry(workflow_run_id)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/retry", path_params: { "workflowRunID" => workflow_run_id })
      end

      # List steps in a workflow run.
      # GET /v1/workflow-runs/{workflowRunID}/steps
      def list_steps(workflow_run_id)
        _request(:get, "/v1/workflow-runs/{workflowRunID}/steps", path_params: { "workflowRunID" => workflow_run_id })
      end

      # Approve a workflow step.
      # POST /v1/workflow-runs/{workflowRunID}/steps/{stepRef}/approve
      def approve_step(workflow_run_id, step_ref, body)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/approve", path_params: { "workflowRunID" => workflow_run_id, "stepRef" => step_ref }, body: body)
      end

      # Retry a workflow step.
      # POST /v1/workflow-runs/{workflowRunID}/steps/{stepRef}/retry
      def retry_step(workflow_run_id, step_ref)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/retry", path_params: { "workflowRunID" => workflow_run_id, "stepRef" => step_ref })
      end

      # Skip a workflow step.
      # POST /v1/workflow-runs/{workflowRunID}/steps/{stepRef}/skip
      def skip_step(workflow_run_id, step_ref)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/skip", path_params: { "workflowRunID" => workflow_run_id, "stepRef" => step_ref })
      end

      # Force-complete a workflow step.
      # POST /v1/workflow-runs/{workflowRunID}/steps/{stepRef}/force-complete
      def force_complete_step(workflow_run_id, step_ref, body)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/force-complete", path_params: { "workflowRunID" => workflow_run_id, "stepRef" => step_ref }, body: body)
      end

      # Replay the subtree starting from a workflow step.
      # POST /v1/workflow-runs/{workflowRunID}/steps/{stepRef}/replay-subtree
      def replay_subtree_step(workflow_run_id, step_ref)
        _request(:post, "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/replay-subtree", path_params: { "workflowRunID" => workflow_run_id, "stepRef" => step_ref })
      end

      # Bulk-cancel workflow runs.
      # POST /v1/workflow-runs/bulk-cancel
      def bulk_cancel(body)
        _request(:post, "/v1/workflow-runs/bulk-cancel", body: body)
      end

      # Bulk-replay workflow runs.
      # POST /v1/workflow-runs/bulk-replay
      def bulk_replay(body)
        _request(:post, "/v1/workflow-runs/bulk-replay", body: body)
      end
    end
  end
end

File: packages/python-sdk/strait/operations/workflow_runs.py
"""Workflow run management operations."""

from __future__ import annotations

from typing import Any

from strait.operations._base import AsyncBaseService, BaseService


class WorkflowRunsService(BaseService):
    def list(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/workflow-runs", query=query)

    def get(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/workflow-runs/{workflowRunID}",
            path_params={"workflowRunID": workflow_run_id},
        )

    def delete(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/workflow-runs/{workflowRunID}",
            path_params={"workflowRunID": workflow_run_id},
        )

    def pause(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/pause",
            path_params={"workflowRunID": workflow_run_id},
        )

    def resume(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/resume",
            path_params={"workflowRunID": workflow_run_id},
        )

    def retry(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/retry",
            path_params={"workflowRunID": workflow_run_id},
        )

    def list_steps(self, workflow_run_id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/workflow-runs/{workflowRunID}/steps",
            path_params={"workflowRunID": workflow_run_id},
        )

    def approve_step(
        self, workflow_run_id: str, step_ref: str, body: Any,
    ) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/approve",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
            body=body,
        )

    def retry_step(self, workflow_run_id: str, step_ref: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/retry",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    def skip_step(self, workflow_run_id: str, step_ref: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/skip",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    def force_complete_step(
        self, workflow_run_id: str, step_ref: str, body: Any,
    ) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/force-complete",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
            body=body,
        )

    def replay_subtree_step(self, workflow_run_id: str, step_ref: str) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/replay-subtree",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    def bulk_cancel(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/workflow-runs/bulk-cancel", body=body)

    def bulk_replay(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/workflow-runs/bulk-replay", body=body)


class AsyncWorkflowRunsService(AsyncBaseService):
    async def list(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return await self._request("GET", "/v1/workflow-runs", query=query)

    async def get(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/workflow-runs/{workflowRunID}",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def delete(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/workflow-runs/{workflowRunID}",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def pause(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/pause",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def resume(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/resume",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def retry(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/retry",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def list_steps(self, workflow_run_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/workflow-runs/{workflowRunID}/steps",
            path_params={"workflowRunID": workflow_run_id},
        )

    async def approve_step(
        self, workflow_run_id: str, step_ref: str, body: Any,
    ) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/approve",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
            body=body,
        )

    async def retry_step(self, workflow_run_id: str, step_ref: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/retry",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    async def skip_step(self, workflow_run_id: str, step_ref: str) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/skip",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    async def force_complete_step(
        self, workflow_run_id: str, step_ref: str, body: Any,
    ) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/force-complete",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
            body=body,
        )

    async def replay_subtree_step(
        self, workflow_run_id: str, step_ref: str,
    ) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/workflow-runs/{workflowRunID}/steps/{stepRef}/replay-subtree",
            path_params={"workflowRunID": workflow_run_id, "stepRef": step_ref},
        )

    async def bulk_cancel(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/workflow-runs/bulk-cancel", body=body)

    async def bulk_replay(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/workflow-runs/bulk-replay", body=body)
